cg/meta/report/field_validators.py

```diff
@@ -0,0 +1,56 @@
+"""Report field validation helper"""
+
+from cg.models.report.report import ReportModel
+
+
+def update_missing_data_dict(missing_data: dict, source: str, field: str, label: str = None):
+    """
+    Populates a specific dictionary with the missing/empty fields.
+    If the new entry is sample/application dependant field, the ID/TAG is also stored to uniquely identify it.
+    """
+
+    if source in missing_data.keys() and label:
+
+        if label in missing_data[source] and isinstance(missing_data[source][label], list):
+            missing_data[source][label].append(field)
+        elif label in missing_data[source]:
+            missing_data[source][label] = [missing_data[source][label], field]
+        else:
+            missing_data[source][label] = field
+    elif source in missing_data.keys() and not label:
+        if isinstance(missing_data[source], list):
+            missing_data[source].append(field)
+        else:
+            missing_data[source] = [missing_data[source], field]
+    else:
+        missing_data.update({source: {label: field} if label else field})
```

Comment:
I'm sure this is a cleaver piece of code, but I can't really get my head around it and I wonder if it can be structured in a more readable way? Does it really need all those ifs?

Reply:
The reasoning behind this method is that I wanted to create a structured log of the missing data following the structure of the retrieved dictionary with the required fields:
```
{
    "report": REQUIRED_REPORT_FIELDS,
    "customer": REQUIRED_CUSTOMER_FIELDS,
    "case": REQUIRED_CASE_FIELDS,
    "applications": REQUIRED_APPLICATION_FIELDS,
    "data_analysis": REQUIRED_DATA_ANALYSIS_MIP_DNA_FIELDS,
    "samples": REQUIRED_SAMPLE_MIP_DNA_FIELDS,
    "methods": REQUIRED_SAMPLE_METHODS_FIELDS,
    "timestamp": REQUIRED_SAMPLE_TIMESTAMPS_FIELDS,
    "metadata": REQUIRED_SAMPLE_METADATA_MIP_DNA_FIELDS,
}
```

Therefore, if a field is missing it will be appended to a dictionary following the above structure and creating a nested dictionary. I have also added a sample/application ID/tag to identify uniquely the missing fields.

And this is the output after appending some mock data:
```
{
    "report": ["accredited", "version"],
    "metadata": {"sample_1": "million_read_pairs", "sample_2": ["duplicates", "pct_10x"]},
    "customer": "name",
    "applications": {"app_1": "tag"},
}
```

Another way of addressing this could be printing directly each of the missing fields individually, which will allow us to remove this method completely. However, the printed log won't be structured as in the example above.

@patrikgrenfeldt Tell me if you would prefer this solution or if you have a better idea to approach this.

And thank you for your reviews!

Reply:
Addressed in [65183d5](https://github.com/Clinical-Genomics/cg/pull/1438/commits/65183d5a915203911b484d7a38b157fa55095118)